ref/src/sw_core.F90

```diff
@@ -1490,8 +1527,8 @@ subroutine write_state(filename)
     call define_dim(ncFileID, "nj", jed-jsd+1, njDimID)
 
     ! Define the i+1, j+1 dimensions
-    call define_dim(ncFileID, "nip1", ied-isd+2, nip1DimID)
-    call define_dim(ncFileID, "njp1", jed-jsd+2, njp1DimID)
+    call define_dim(ncFileID, "nip1", ied-isd+2 + interpFactor, nip1DimID)
```

Comment:
This doesn't seem right.  The size of the dimensions is increased not by `interpFactor` but by that value times the number of times it was applied.

Reply:
@christopherwharrop-noaa What do you mean by: "but by that value times the number of times it was applied"?

Reply:
If you have a `3 x 3` grid, your dimension variables are `3` and `3`.  If you then apply an interpolation factor of 2, you get a `7 x 7` grid.  That is, the dimension of each side doesn't increase by `2`, it increases by `4` because the interpolation factor is applied two times.

So, in this case, `nip1` is not increased by `interpFactor`, it would be increased by interpFactor multiplied by the number of times `interpFactor` points were inserted.

But, now that I think about this more, the whole thing seems moot.  This change shouldn't be made at all because the `write_state` should write the current state, and all of the dimensions should already have been updated to new values.  So the original lines should be restored.

Reply:
@christopherwharrop-noaa The variable divg_d is one of the ied+1, jed+1 variables.  Here is where it is allocated under interpFactor>0.

    odims(1) = isd
    odims(2) = ied + 1
    odims(3) = jsd
    odims(4) = jed + 1
    odims(5) = 1
    odims(6) = npz
    call interpolateCalculateSpace3D(odims, interpFactor, idims)
    allocate (new_divg_d(idims(1):idims(2), idims(3):idims(4), idims(5):idims(6)))
    call interpolateArray3D(divg_d, odims, new_divg_d, idims, interpFactor)

Note that divg_d's new dimensions are calculated with knowledge of interpFactor.  And we know that ied and jed are updated with their new values after the interpolation has occurred.  

Now we are trying to write out divg_d and we know the new values of ied and jed.  But we want the new values of ied+1 and jed+1 under interpFactor.  So we can not use isd+1 and jsd+1, we need to use isd+interpFactor and jsd+interpFactor.  Make sense?

Reply:
No. That doesn't make sense to me.  How are all those points between `ied` and `ied+1` being accounted for in the computations?  I don't think this will work correctly.  The calculations rely on `divg_d` being one element larger.  That means the new value of `ied+1` and `jed+1` has to be one bigger than the new value of `ied` and `jed`.  I still don't think this is correct.

Reply:
I put this comment in the wrong location.  This is a comment to the above.

@christopherwharrop-noaa  Re: "How are all those points between ied and ied+1 being accounted for in the computations?"

The subroutine: write_state is called after the interpolation. So ied and jed have already been expanded by interpFactor. But the "+1" has not been expanded. That is why write_state needs to be told about interpFactor so the dimensions of divg_d are calculated correctly.

Reply:
I added the following print lines before the definition of nip1DimID.  This is inserted code in sw_core.F90.

    ! Define the i+1, j+1 dimensions
    print *,"ied, jed, ied-isd+2, jed-jsd+2=",ied, jed, ied-isd+2, jed-jsd+2
    print *,"shape(divg_d)=", shape(divg_d)
    call define_dim(ncFileID, "nip1", ied-isd+2 + interpFactor, nip1DimID)
    call define_dim(ncFileID, "njp1", jed-jsd+2 + interpFactor, njp1DimID)

This this is the output:

[Lynd.Stringer@hfe11 exe]$ ./c_sw c_sw_12x24_3.nl
runnable: found another runnable range starting at core=40. lo_out=0 hi_out=0
Thread 0 on hfe11. (Runnable range: lo=0 hi=0)
runnable: found another runnable range starting at core=41. lo_out=1 hi_out=1
Thread 1 on hfe11. (Runnable range: lo=1 hi=1)
runnable: found another runnable range starting at core=42. lo_out=2 hi_out=2
Thread 2 on hfe11. (Runnable range: lo=2 hi=2)
runnable: found another runnable range starting at core=43. lo_out=3 hi_out=3
Thread 3 on hfe11. (Runnable range: lo=3 hi=3)
 ied, jed, ied-isd+2, jed-jsd+2=          66         114          70         118
 shape(divg_d)=          73         121         127

Note that the shape of divg_d is  (73, 121,127).  But ied-isd+2 is 70, our definition of the first dimension.  If we do not add interpFactor, when we try to write divg_d to the netcdf file, we do not have the correct dimensions and we get the an error.  That is why interpFactor is added to write_state.

Reply:
@LyndStringer - I think I see what you're getting at, but it still doesn't make sense.  In `allocate_state()` there is this line:

```
    allocate(divg_d (isd:ied+1, jsd:jed+1, npz))
```

Which allocates `divg_d` to be shape `isd:ied+1, jsd:jed+1`.  It is using the updated values of `ied` and `jed` because you are calling `allocate_state()` in the interpolation routine after the values of `ied` and `jed` have been updated.  So, how is it possible that the shape of `divg_d` is something else?

I still feel like something is fundamentally wrong here.  All the calculations in the code where there is a `+1` indexing rely on those arrays to be one bigger, not `1 + interpFactor` bigger.  Moreover, it looks to me like the temporary arrays in the interpolate routine are allocated inconsistently with how the state variables are finally allocated by `allocate_state()` at the end of the routine.  I don't see how those match up due to the allocation discrepancy caused by the interpolation between the `ied` and `ied+1` index.

Reply:
Here's the timing.  
1) allocate_state is called, divg_d is allocated (just like you mentioned above)
2) if interpFactor is > 0, divg_d is deallocated and reallocated with new and larger dimensions
      * the larger dimensions are calculated:
      
          odims(1) = isd
          odims(2) = ied + 1
          odims(3) = jsd
          odims(4) = jed + 1
          odims(5) = 1
          odims(6) = npz
          call interpolateCalculateSpace3D(odims, interpFactor, idims)
          allocate (new_divg_d(idims(1):idims(2), idims(3):idims(4), idims(5):idims(6)))
          call interpolateArray3D(divg_d, odims, new_divg_d, idims, interpFactor)

         This is the important point, divg_d is expanded such that just adding 1 to the new ied 
         will not work because the old ied+1 was expanded

      * see line 1991 in sw_core.F90, this is where divg_d (and others) is re-defined
3) if interpFactor is 0, no interpolation is done
4) write_state is called

I was not happy adding interpFactor to write_state, but just adding +1 was not working, netcdf was 
declaring an error because the dimensions of divg_d and the netcdf dimensions were not 
identical.  If there is a better way, I'm all for it.  But I do not have a better solution currently.

Reply:
I don't see how that sequence could be correct.  Here is an excerpt from the new `interpolate_state` routine:

```
    odims(1) = isd
    odims(2) = ied + 1
    odims(3) = jsd
    odims(4) = jed + 1
    odims(5) = 1
    odims(6) = npz
    call interpolateCalculateSpace3D(odims, interpFactor, idims)
    allocate (new_divg_d(idims(1):idims(2), idims(3):idims(4), idims(5):idims(6)))
    call interpolateArray3D(divg_d, odims, new_divg_d, idims, interpFactor)

    ! Exchange the old dimensions to the new dimensions.
    ied = new_ied
    jed = new_jed
    is = 1
    ie = new_ied - 3
    js = 1
    je = new_jed - 3
    nord = 1
    npx = ie + 1
    npy = je + 1
    npz = 127

    call deallocate_state()
    call allocate_state()
```

The state arrays are allocated AFTER the interpolation is done and new values for the dimension variables have been set.

Reply:
Appears the assignment into divg_d changes the subscripts, that is to say, divg_d is redefined:

    print *,"shape(divg_d)=",shape(divg_d)
    divg_d = new_divg_d
    print *,"shape(divg_d)=",shape(divg_d)

 shape(divg_d)=          70         118         127
 shape(divg_d)=          73         121         127

Reply:
Ugh... That is not good.  I didn't think of that.  Try:

```
divg_d(:,:,:) = new_divg_d(:,:,:)
```

instead of

```
divg_d = new_divg_d
```

Does that also change the dimensions?  In any case, I like using the colons to make it more explicit about what is happening in the assignment.

Reply:
I would suggest a different approach, delete the two lines above the assignment section which are not needed.  They are:

    call deallocate_state()
    call allocate_state()

and add a comment to the assignment section noting that the assignments not only redefines the values but also the subscript ranges are redefined.

Reply:
I respectfully disagree with that idea.  It's not what we need to have happen.  The issue is that the number of elements in the i, and j directions of `divg_d` needs to be ONE bigger, not `interpFactor + 1` bigger.

Reply:
What I'm saying is that the way the new dimensions are being calculated is not correct for how the code works.  After new values for `ied` and `jed` are computed, the allocation of the data needs to be the way it is allocated in the existing `allocate_state()` routine.  Otherwise, the computation doesn't make any sense because of the way the code indexes elements of the arrays and expects them to be sized.

Reply:
To put it another way, the interpolation has to work without making any changes to the `allocate_state()` or `write_state()` routines.

Reply:
I have an idea that might be a very simple way to solve this problem with minimal effort.  All the `+1`s are causing problems because of the introduction of new points between the `ied` and `jed` and the `ied+1` and `jed+1`.  I think the solution is actually very simple.  We only want the data from `isd:ied+1, jsd:jed+1`  We don't want/need the data from `ied+2:ied+1+interpFactor` and `jed:+2:jed+1+interpFactor` because it isn't used.  So, what if we keep the interpolation calls the same because that's what's needed to get the interpolated values we want.  But... when we copy the data from the `*_new` arrays (NOTE: this is AFTER we've computed new values for `ied` and `jed` etc), we only copy the range of values from `isd:ied+1` rather than all the values (i.e. `isd:ied+1+interpFactor`).  Then, the dimensions are consistent with the code and allocation and I/O.  We just throw away those extras because they're useless to us, anyway.

Reply:
@LyndStringer - Forgot to tag you.  Please see note just above.

Reply:
In case it wasn't clear, what I'm proposing is to call `allocate_state()` after new values for `ied` `jed` etc are computed by the interpolation routines.  And then, in the final assignment, copy the explicit indices that are needed for each one.  For example:

```
divg_d(isd:ied+1, jsd:jed+1, :) = new_divg_d(isd:ied+1, jsd:jed+1, :)
```

All of the assignments would be like that so that we are not copying indices > `ied+1` and `jed+1`.  This keeps the dimensions correct, and uses all the interpolated values between `isd` and `ied+1` (and between `jsd` and `jed+1`) which are the ones actually used in the computation.

Reply:
Done.